Make ephemeral resource creation optional

ayoubbergaoui opened the pull request:
# Make ephemeral resource creation optional

### Description
This pull request introduces a new configuration option in the Lockable Resources Plugin to control the creation of ephemeral resources. The goal is to give administrators more control over resource management by allowing them to prevent automatic creation of ephemeral resources.

### Changes
- Added a new boolean field `allowEphemeralResources` in `LockableResourcesManager` (default: true)
- Updated the system configuration UI with a new checkbox option "Allow Ephemeral Resources"
- Enhanced error handling with clear user guidance:
  * For missing resources: Lists which resources are missing and how to fix it
  * For invalid labels: Shows which label doesn't match and possible solutions
- Updated documentation in README.md with the new feature
- Added configuration-as-code (JCasC) support

### Implementation Details
The changes span across several files:
- `LockableResourcesManager.java`: Core logic for the new feature
- `LockableResourcesStruct.java`: Enhanced validation for resources and labels
- `LockableResourcesQueueTaskDispatcher.java`: Improved error messages
- `config.jelly` and `config.properties`: UI changes
- `README.md`: Documentation updates

### Testing done
- Verified functionality with ephemeral resources enabled (default behavior)
- Tested with ephemeral resources disabled:
  * Missing resource case
  * Invalid label case
  * Multiple missing resources case
- Validated error messages clarity
- Confirmed JCasC compatibility

### Proposed upgrade guidelines
N/A - This change is backward compatible. The feature is enabled by default, maintaining existing behavior.

### Localizations
New strings added in `config.properties`:
- [x] English

### Notes
This change improves the plugin's flexibility by making ephemeral resource creation optional while maintaining backward compatibility.

mPokornyETM commented:
@ayoubbergaoui  Did this change fixed  #651 ??

ayoubbergaoui commented:
@mPokornyETM  yes! it's the same need (even if I haven't seen that there have been other people who need this)

ayoubbergaoui commented:
### Test Results: Non-Existing Resource/Label with Ephemeral Resources Disabled

Just sharing the results of the tests I ran locally.

When **ephemeral resources are disabled** and we specify a **resource name that does not exist** in the list of lockable resources, this is what happens:

![Resource Not Found](https://i.imghippo.com/files/eI2599Lk.png)  
![Resource Not Found - Details](https://i.imghippo.com/files/wDo7069jD.png)

Similarly, when specifying a **label name that does not exist**, we get:

![Label Not Found](https://i.imghippo.com/files/yJt4953uKc.png)  
![Label Not Found - Details](https://i.imghippo.com/files/gD4423dE.png)

ayoubbergaoui commented:
@mPokornyETM  can you review / validate the PR please?

mPokornyETM commented:
> @mPokornyETM
> 
> ### Test Results: Non-Existing Resource/Label with Ephemeral Resources Disabled
> Just sharing the results of the tests I ran locally.
> 
> When **ephemeral resources are disabled** and we specify a **resource name that does not exist** in the list of lockable resources, this is what happens:
> 
> ![Resource Not Found](https://camo.githubusercontent.com/451180e0bba36986728f938800cdd0f8b95898692041d036fdc0b73aa9b439d3/68747470733a2f2f692e696d67686970706f2e636f6d2f66696c65732f6549323539394c6b2e706e67) ![Resource Not Found - Details](https://camo.githubusercontent.com/0b4dd8cf9d69fc35ffaa17350fb9f86b606f600b3a703251e291f737627d2344/68747470733a2f2f692e696d67686970706f2e636f6d2f66696c65732f77446f373036396a442e706e67)
> 
> Similarly, when specifying a **label name that does not exist**, we get:
> 
> ![Label Not Found](https://camo.githubusercontent.com/438df081d20a4aa9f0d05e265143709ebbf9fdd8ff2e77d33351d0d8e3b1c860/68747470733a2f2f692e696d67686970706f2e636f6d2f66696c65732f794a7434393533754b632e706e67) ![Label Not Found - Details](https://camo.githubusercontent.com/ba927524e47507835d6e175be34e736c3801ef7e0109c49713b417044f50e924/68747470733a2f2f692e696d67686970706f2e636f6d2f66696c65732f67443434323364452e706e67)

I mean automated tests. Sorry, I am very strict in this plugin, because it is used in many instances. It is complex and all touches are dangerous. Therefory I propose autimated tests for every code change, when possible

ayoubbergaoui commented:
@mPokornyETM I just updated my commit with a set of tests that cover all my changes.

So all the lines I added are covered now by a test!

If you have time, can you validate the PR please?

Thanks in advance,

ThomasFaivre reviewed (changes requested):
Hi all,

Small change in documentation is needed (see review comment).

I feel like there are some cases that are not handled completely/correctly:

> src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java
```java
    private boolean proceedNextContext() {
    ...
            List<LockableResource> requiredResourceForNextContext =
                this.fromNames(nextContext.candidates, /*create un-existent resources */ true);
```
I don't know what this proceedNextContext method is, but the `true` is ignored due to the check for the option in createResource. It feels confusing.

Same here:
```java
    public List<LockableResource> getAvailableResources(
    ...
                available = fromNames(
                        getResourcesNames(requiredResources.required), /*create un-existent resources */ true);
```

These inconsistencies make me think that some tests are missing:

> src/test/java/org/jenkins/plugins/lockableresources/LockStepWithRestartTest.java
```java
    public void chaosOnRestart() throws Throwable {
    ...
                            + "    lock('ephemeral_' + env.BUILD_NUMBER + '_' + resourceName) {\n"
                            + "      lock(resourceName) {semaphore semId}\n"
```
> src/test/java/org/jenkins/plugins/lockableresources/PressureTest.java
```java
    public void pressure(final int resourcesCount) throws Exception {
    ...
                + "    lock('resource_ephemeral_' + stageName) {\n"
                // + "      echo \"*** locked resource_ephemeral_\" + stageName\n"
                + "    }\n"
```

The tests passed with current code because the ephemeral option is enabled by default.

Same idea with Configuration-as-Code: shouldn't the value of allowEphemeralResources be checked when false?

I'll let the maintainers way in on this.
Otherwise, LGTM (Disclaimer: I'm biased)

mPokornyETM commented:
#651 was fixed
